File: windows/Gutenkern.Core/ResultLayout.cs
namespace Gutenkern;

public enum ResultLayoutMode
{
    Row,
    Column
}

public sealed class ResultGroup
{
    public string GroupId { get; }
    public KerningGroup Category { get; }
    public IReadOnlyList<PairLine> Pairs { get; }
    public IReadOnlyList<string> Keys { get; }

    public string CopyText => string.Join("  ", Pairs.Select(pair => pair.Display));

    public ResultGroup(string groupId, KerningGroup category, IReadOnlyList<PairLine> pairs)
    {
        GroupId = groupId;
        Category = category;
        Pairs = pairs;
        Keys = pairs.Select(pair => pair.Key).ToList();
    }
}

public sealed class ResultCategory
{
    public KerningGroup Group { get; }
    public IReadOnlyList<ResultGroup> Blocks { get; }

    public ResultCategory(KerningGroup group, IReadOnlyList<ResultGroup> blocks)
    {
        Group = group;
        Blocks = blocks;
    }
}

public sealed class ResultToken
{
    public string Key { get; }
    public string Display { get; }
    public string GroupId { get; }
    public KerningGroup Category { get; }
    public int Utf16Start { get; }
    public int Utf16Length { get; }
    public int Utf16End => Utf16Start + Utf16Length;

    public ResultToken(
        string key,
        string display,
        string groupId,
        KerningGroup category,
        int utf16Start,
        int utf16Length)
    {
        Key = key;
        Display = display;
        GroupId = groupId;
        Category = category;
        Utf16Start = utf16Start;
        Utf16Length = utf16Length;
    }
}

public sealed class ResultLayout
{
    public static ResultLayout Empty { get; } = new(
        "",
        [],
        new Dictionary<KerningGroup, int>(),
        []);

    public string Text { get; }
    public string ViewText { get; }
    public IReadOnlyList<ResultToken> Tokens { get; }
    public IReadOnlyDictionary<KerningGroup, int> CategoryStarts { get; }
    public IReadOnlyList<ResultCategory> Categories { get; }

    public ResultLayout(
        string text,
        IReadOnlyList<ResultToken> tokens,
        IReadOnlyDictionary<KerningGroup, int> categoryStarts,
        IReadOnlyList<ResultCategory> categories)
    {
        Text = text;
        Tokens = tokens;
        CategoryStarts = categoryStarts;
        Categories = categories;
        ViewText = TokenGlue.ViewText(text, tokens);
    }

    public bool IsEmpty => Text.Length == 0;

    public bool HasCategory(KerningGroup group) => CategoryStarts.ContainsKey(group);

    public (int Done, int Total) Progress(KerningGroup group, IReadOnlySet<string> done) =>
        ProgressByCategory(done).GetValueOrDefault(group);

    public Dictionary<KerningGroup, (int Done, int Total)> ProgressByCategory(IReadOnlySet<string> done)
    {
        var totals = new Dictionary<KerningGroup, (int Done, int Total)>();
        foreach (var token in Tokens)
        {
            totals.TryGetValue(token.Category, out var entry);
            entry.Total++;
            if (done.Contains(token.Key))
            {
                entry.Done++;
            }

            totals[token.Category] = entry;
        }

        return totals;
    }

    public static ResultLayout Build(
        IReadOnlyList<RecipeSection> sections,
        ResultLayoutMode mode = ResultLayoutMode.Column)
    {
        var recipeByLine = KerningPlan.Recipes.ToDictionary(recipe => recipe.Line, recipe => recipe);
        var text = new System.Text.StringBuilder();
        var tokens = new List<ResultToken>();
        var categoryStarts = new Dictionary<KerningGroup, int>();
        var groupCounters = new Dictionary<KerningGroup, int>();
        var categories = new List<ResultCategory>();
        KerningGroup? currentCategory = null;
        var currentBlocks = new List<ResultGroup>();
        var pairSeparator = mode == ResultLayoutMode.Row ? "  " : "\n";

        void FlushCategory()
        {
            if (currentCategory is { } group && currentBlocks.Count > 0)
            {
                categories.Add(new ResultCategory(group, currentBlocks.ToList()));
                currentBlocks.Clear();
            }
        }

        for (var sectionIndex = 0; sectionIndex < sections.Count; sectionIndex++)
        {
            if (sectionIndex > 0)
            {
                text.Append("\n\n\n");
            }

            var section = sections[sectionIndex];
            if (!recipeByLine.TryGetValue(section.Recipe, out var recipe))
            {
                continue;
            }

            var category = recipe.Left;
            if (currentCategory != category)
            {
                FlushCategory();
                currentCategory = category;
            }

            for (var blockIndex = 0; blockIndex < section.PairGroups.Count; blockIndex++)
            {
                if (blockIndex > 0)
                {
                    text.Append("\n\n");
                }

                if (!categoryStarts.ContainsKey(category))
                {
                    categoryStarts[category] = text.Length;
                }

                var index = groupCounters.GetValueOrDefault(category);
                groupCounters[category] = index + 1;
                var groupId = $"{KerningPlan.Code(category)}-{index}";
                var pairs = section.PairGroups[blockIndex];
                for (var pairIndex = 0; pairIndex < pairs.Count; pairIndex++)
                {
                    if (pairIndex > 0)
                    {
                        text.Append(pairSeparator);
                    }

                    var pair = pairs[pairIndex];
                    var start = text.Length;
                    text.Append(pair.Display);
                    tokens.Add(new ResultToken(
                        pair.Key,
                        pair.Display,
                        groupId,
                        category,
                        start,
                        pair.Display.Length));
                }

                currentBlocks.Add(new ResultGroup(groupId, category, pairs));
            }
        }

        FlushCategory();
        return new ResultLayout(text.ToString(), tokens, categoryStarts, categories);
    }

    public ResultToken? TokenAtUtf16(int index)
    {
        foreach (var token in Tokens)
        {
            if (index >= token.Utf16Start && index < token.Utf16End)
            {
                return token;
            }
        }

        return NearestToken(index);
    }

    public ResultToken? NearestToken(int index)
    {
        ResultToken? best = null;
        var bestDistance = int.MaxValue;
        foreach (var token in Tokens)
        {
            int distance;
            if (index < token.Utf16Start)
            {
                distance = token.Utf16Start - index;
            }
            else if (index >= token.Utf16End)
            {
                distance = index - token.Utf16End + 1;
            }
            else
            {
                return token;
            }

            if (distance < bestDistance)
            {
                bestDistance = distance;
                best = token;
            }
        }

        return best;
    }

    public IReadOnlyList<ResultToken> TokensInUtf16(int start, int length)
    {
        if (length <= 0)
        {
            var token = TokenAtUtf16(start);
            return token is null ? [] : [token];
        }

        var end = start + length;
        return Tokens.Where(token => token.Utf16Start < end && token.Utf16End > start).ToList();
    }

    public IReadOnlyList<string> GroupKeys(string groupId) =>
        Tokens.Where(token => token.GroupId == groupId).Select(token => token.Key).ToList();

    public (int Start, int Length)? Utf16Range(string groupId)
    {
        ResultToken? first = null;
        ResultToken? last = null;
        foreach (var token in Tokens)
        {
            if (token.GroupId != groupId)
            {
                continue;
            }

            first ??= token;
            last = token;
        }

        if (first is null || last is null)
        {
            return null;
        }

        return (first.Utf16Start, last.Utf16End - first.Utf16Start);
    }

    public KerningGroup? CategoryAtUtf16(int index)
    {
        KerningGroup? match = null;
        var bestStart = -1;
        foreach (var (group, start) in CategoryStarts)
        {
            if (start <= index && start >= bestStart)
            {
                bestStart = start;
                match = group;
            }
        }

        return match;
    }

    public Dictionary<string, HashSet<string>> KeysByGroup()
    {
        var keys = new Dictionary<string, HashSet<string>>();
        foreach (var token in Tokens)
        {
            if (!keys.TryGetValue(token.GroupId, out var group))
            {
                group = [];
                keys[token.GroupId] = group;
            }

            group.Add(token.Key);
        }

        return keys;
    }

    public bool HasMixedGroups(KerningMarkState state) =>
        KeysByGroup().Values.Any(keys => state.GroupMark(keys.ToList()) == GroupMark.MixedDone);
}

public static class NewUnkernedNotice
{
    public static (HashSet<string> WarningGroupIds, HashSet<string> NewKeys) Update(
        IReadOnlyDictionary<string, HashSet<string>> previous,
        IReadOnlyDictionary<string, HashSet<string>> current,
        IReadOnlySet<string> done,
        IReadOnlySet<string> warningGroupIds,
        IReadOnlySet<string> newKeys)
    {
        var warning = new HashSet<string>(warningGroupIds);
        var news = new HashSet<string>(newKeys);
        var firstPass = previous.Count == 0;
        foreach (var (groupId, keys) in current)
        {
            if (!IsMixed(keys, done))
            {
                warning.Remove(groupId);
                news.ExceptWith(keys);
                continue;
            }

            if (firstPass)
            {
                continue;
            }

            var before = previous.TryGetValue(groupId, out var prior) ? prior : [];
            var added = keys.Except(before).ToHashSet();
            var hadDone = before.Overlaps(done);
            if (added.Count > 0 && hadDone)
            {
                warning.Add(groupId);
                news.UnionWith(added);
            }
        }

        warning.RemoveWhere(groupId => !current.ContainsKey(groupId));
        var known = current.Values.SelectMany(keys => keys).ToHashSet();
        news.IntersectWith(known);
        return (warning, news);
    }

    private static bool IsMixed(HashSet<string> keys, IReadOnlySet<string> done)
    {
        if (keys.Count == 0)
        {
            return false;
        }

        var doneCount = keys.Count(done.Contains);
        return doneCount > 0 && doneCount < keys.Count;
    }
}

public static class ResultSelectionMarks
{
    public static bool CanStrike(IReadOnlyList<string> keys, KerningMarkState state) =>
        keys.Any(key => state.PairMark(key) == PairMark.Empty);

    public static bool CanUnstrike(IReadOnlyList<string> keys, KerningMarkState state) =>
        keys.Any(key => state.PairMark(key) == PairMark.Done);

    public static KerningMarkState Strike(IReadOnlyList<string> keys, KerningMarkState state) =>
        KerningMarks.MarkDone(keys, state);

    public static KerningMarkState Unstrike(IReadOnlyList<string> keys, KerningMarkState state)
    {
        var next = new HashSet<string>(state.Done);
        foreach (var key in keys)
        {
            next.Remove(key);
        }

        return new KerningMarkState(next);
    }
}

File: macos/Gutenkern/Sources/GutenkernCore/ResultLayout.swift
import Foundation

public struct ResultToken: Equatable, Sendable {
    public let key: String
    public let display: String
    public let groupId: String
    public let category: KerningGroup
    public let utf16Start: Int
    public let utf16Length: Int

    public var utf16End: Int { utf16Start + utf16Length }

    public init(
        key: String,
        display: String,
        groupId: String,
        category: KerningGroup,
        utf16Start: Int,
        utf16Length: Int
    ) {
        self.key = key
        self.display = display
        self.groupId = groupId
        self.category = category
        self.utf16Start = utf16Start
        self.utf16Length = utf16Length
    }
}

public struct ResultGroup: Equatable, Identifiable, Sendable {
    public var id: String { groupId }
    public let groupId: String
    public let category: KerningGroup
    public let pairs: [PairLine]

    public var keys: [String] { pairs.map(\.key) }

    public init(groupId: String, category: KerningGroup, pairs: [PairLine]) {
        self.groupId = groupId
        self.category = category
        self.pairs = pairs
    }
}

public struct ResultCategory: Equatable, Identifiable, Sendable {
    public var id: KerningGroup { group }
    public let group: KerningGroup
    public let blocks: [ResultGroup]

    public init(group: KerningGroup, blocks: [ResultGroup]) {
        self.group = group
        self.blocks = blocks
    }
}

public enum ResultLayoutMode: String, CaseIterable, Sendable {
    case row
    case column
}

public struct ResultLayout: Equatable, Sendable {
    public let text: String
    public let viewText: String
    public let tokens: [ResultToken]
    public let categoryStarts: [KerningGroup: Int]
    public let categories: [ResultCategory]

    public static let empty = ResultLayout(text: "", tokens: [], categoryStarts: [:], categories: [])

    public init(
        text: String,
        tokens: [ResultToken],
        categoryStarts: [KerningGroup: Int],
        categories: [ResultCategory]
    ) {
        self.text = text
        self.tokens = tokens
        self.categoryStarts = categoryStarts
        self.categories = categories
        viewText = TokenGlue.viewText(layoutText: text, tokens: tokens)
    }

    public var isEmpty: Bool { text.isEmpty }

    public func hasCategory(_ group: KerningGroup) -> Bool {
        categoryStarts[group] != nil
    }

    public func progress(for group: KerningGroup, done: Set<String>) -> (done: Int, total: Int) {
        progressByCategory(done: done)[group] ?? (0, 0)
    }

    public func progressByCategory(done: Set<String>) -> [KerningGroup: (done: Int, total: Int)] {
        var totals: [KerningGroup: (done: Int, total: Int)] = [:]
        for token in tokens {
            var entry = totals[token.category, default: (0, 0)]
            entry.total += 1
            if done.contains(token.key) {
                entry.done += 1
            }
            totals[token.category] = entry
        }
        return totals
    }

    public static func build(sections: [RecipeSection], mode: ResultLayoutMode = .column) -> ResultLayout {
        let recipeByLine = Dictionary(uniqueKeysWithValues: KerningPlan.recipes.map { ($0.line, $0) })
        var text = ""
        var tokens: [ResultToken] = []
        var categoryStarts: [KerningGroup: Int] = [:]
        var groupCounters: [KerningGroup: Int] = [:]
        var categories: [ResultCategory] = []
        var currentCategory: KerningGroup?
        var currentBlocks: [ResultGroup] = []
        let pairSeparator = mode == .row ? "  " : "\n"

        func flushCategory() {
            guard let currentCategory, !currentBlocks.isEmpty else {
                return
            }
            categories.append(ResultCategory(group: currentCategory, blocks: currentBlocks))
            currentBlocks = []
        }

        for (sectionIndex, section) in sections.enumerated() {
            if sectionIndex > 0 {
                text += "\n\n\n"
            }
            guard let recipe = recipeByLine[section.recipe] else {
                continue
            }
            let category = recipe.left
            if currentCategory != category {
                flushCategory()
                currentCategory = category
            }
            for (blockIndex, pairs) in section.pairGroups.enumerated() {
                if blockIndex > 0 {
                    text += "\n\n"
                }
                if categoryStarts[category] == nil {
                    categoryStarts[category] = utf16Count(text)
                }
                let index = groupCounters[category, default: 0]
                groupCounters[category] = index + 1
                let groupId = "\(category.rawValue)-\(index)"
                for (pairIndex, pair) in pairs.enumerated() {
                    if pairIndex > 0 {
                        text += pairSeparator
                    }
                    let start = utf16Count(text)
                    let length = utf16Count(pair.display)
                    text += pair.display
                    tokens.append(
                        ResultToken(
                            key: pair.key,
                            display: pair.display,
                            groupId: groupId,
                            category: category,
                            utf16Start: start,
                            utf16Length: length
                        )
                    )
                }
                currentBlocks.append(ResultGroup(groupId: groupId, category: category, pairs: pairs))
            }
        }
        flushCategory()

        return ResultLayout(
            text: text,
            tokens: tokens,
            categoryStarts: categoryStarts,
            categories: categories
        )
    }

    public func token(atUtf16 index: Int) -> ResultToken? {
        for token in tokens {
            if index >= token.utf16Start && index < token.utf16End {
                return token
            }
        }
        return nearestToken(atUtf16: index)
    }

    public func nearestToken(atUtf16 index: Int) -> ResultToken? {
        var best: ResultToken?
        var bestDistance = Int.max
        for token in tokens {
            let distance: Int
            if index < token.utf16Start {
                distance = token.utf16Start - index
            } else if index >= token.utf16End {
                distance = index - token.utf16End + 1
            } else {
                return token
            }
            if distance < bestDistance {
                bestDistance = distance
                best = token
            }
        }
        return best
    }

    public func tokens(utf16Start start: Int, length: Int) -> [ResultToken] {
        if length <= 0 {
            if let token = token(atUtf16: start) {
                return [token]
            }
            return []
        }
        let end = start + length
        return tokens.filter { token in
            token.utf16Start < end && token.utf16End > start
        }
    }

    public func groupKeys(forGroupId groupId: String) -> [String] {
        tokens.filter { $0.groupId == groupId }.map(\.key)
    }

    public func utf16Range(forGroupId groupId: String) -> (start: Int, length: Int)? {
        var first: ResultToken?
        var last: ResultToken?
        for token in tokens where token.groupId == groupId {
            if first == nil {
                first = token
            }
            last = token
        }
        guard let first, let last else {
            return nil
        }
        return (first.utf16Start, last.utf16End - first.utf16Start)
    }

    public func category(atUtf16 index: Int) -> KerningGroup? {
        var match: KerningGroup?
        var bestStart = -1
        for (group, start) in categoryStarts {
            if start <= index, start >= bestStart {
                bestStart = start
                match = group
            }
        }
        return match
    }

    public func keysByGroup() -> [String: Set<String>] {
        var keys: [String: Set<String>] = [:]
        for token in tokens {
            keys[token.groupId, default: []].insert(token.key)
        }
        return keys
    }
}

public enum NewUnkernedNotice {
    public static func update(
        previous: [String: Set<String>],
        current: [String: Set<String>],
        done: Set<String>,
        warningGroupIds: Set<String>,
        newKeys: Set<String>
    ) -> (warningGroupIds: Set<String>, newKeys: Set<String>) {
        var warning = warningGroupIds
        var news = newKeys
        let firstPass = previous.isEmpty
        for (groupId, keys) in current {
            if !isMixed(keys: keys, done: done) {
                warning.remove(groupId)
                news.subtract(keys)
                continue
            }
            if firstPass {
                continue
            }
            let before = previous[groupId] ?? []
            let added = keys.subtracting(before)
            let hadDone = !before.intersection(done).isEmpty
            if !added.isEmpty, hadDone {
                warning.insert(groupId)
                news.formUnion(added)
            }
        }
        warning = warning.filter { current[$0] != nil }
        let known = Set(current.values.flatMap { $0 })
        news = news.intersection(known)
        return (warning, news)
    }

    private static func isMixed(keys: Set<String>, done: Set<String>) -> Bool {
        guard !keys.isEmpty else {
            return false
        }
        let doneCount = keys.reduce(into: 0) { count, key in
            if done.contains(key) {
                count += 1
            }
        }
        return doneCount > 0 && doneCount < keys.count
    }
}

public enum ResultSelectionMarks {
    public static func canStrike(keys: [String], state: KerningMarkState) -> Bool {
        keys.contains { state.pairMark($0) == .empty }
    }

    public static func canUnstrike(keys: [String], state: KerningMarkState) -> Bool {
        keys.contains { state.pairMark($0) == .done }
    }

    public static func strike(keys: [String], state: KerningMarkState) -> KerningMarkState {
        KerningMarks.markDone(keys: keys, state: state)
    }

    public static func unstrike(keys: [String], state: KerningMarkState) -> KerningMarkState {
        var next = state
        for key in keys {
            next.done.remove(key)
        }
        return next
    }
}

private func utf16Count(_ string: String) -> Int {
    (string as NSString).length
}
